### PR TITLE
Misc refactorings

### PR DESCRIPTION
**Pull request by theobat**
Minor refactorings and additions for component based builds

### CODE REVIEW

**Comment by theobat** [comment]
This is currently failing due to a bug in ghc.
It's arguably related to 9.4.* and should not be there in 9.6

see 
https://gitlab.haskell.org/ghc/ghc/-/issues/23972

**Comment by mpilgrem** [comment]
@theobat, I'm currently 'almost there' in moving Stack on to GHC 9.6.3 (see #6384).

**Comment by mpilgrem** [comment]
@theobat, I rebased on master (and a GHC 9.6.3-based Stack code base).

**Comment by mpilgrem** [comment]
I am going to close and reopen the pull request, with the intent of forcing the CI to run.

**Comment by mpilgrem** [comment]
I've taken the opportunity to make use of helper function `fromPackageName` consistently (both in these changes and elsewhere).

File: src/Stack/Types/NamedComponent.hs [hunk 1]
```diff
@@ -119,3 +132,4 @@ splitComponents =
   go s e t b (CExe x:xs) = go s (e . (x:)) t b xs
   go s e t b (CTest x:xs) = go s e (t . (x:)) b xs
   go s e t b (CBench x:xs) = go s e t (b . (x:)) xs
+  go s e t b x = go s e t b x
```
Comment (mpilgrem) [inline src/Stack/Types/NamedComponent.hs:135]: @theobat, as the righthand side is the same as the lefthand side, how will this ever terminate? Presumably `x` would be `(CFlib x : xs)`, as the other patterns are catered for. I wondered if this was intended to be `go s e t b (CFlib x : xs) = go s e t b xs`?

**Review by mpilgrem — CHANGES_REQUESTED** [review]
I had only one substantive observation.

File: src/Stack/Types/NamedComponent.hs (hunk 1 above)
Comment (theobat) [reply src/Stack/Types/NamedComponent.hs:135]: that's right, good catch, I'll fix it

**Comment by mpilgrem** [comment]
In the force-pushing, my reformatting commit has been lost. I'll see if I can reapply it from my local version this evening.

**Comment by theobat** [comment]
@mpilgrem I'm sorry I lost your reformatting I'll try to pull before doing my rebase next time. This last force push should have repaired the broken tests, I only provide the missing info to ghci when there's actually a dependance on a sublibrary.

So, it turns out this MR also contains the fix for the `stack ghci my-project:lib` example that failed before

**Pull request merged**